File: web/src/entities/coin.ts
import { Direction } from '../constants.js';
import { Vec2, ImmutableVec2, type ReadonlyVec2 } from '../math.js';
import { sincos } from '../utils.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';
import type { Room } from '../room.js';

const COIN_SIZE = 12;

const enum CoinType { DYNAMIC, STATIC }

export class Coin extends Entity {
  private readonly animationCoin: Animation;
  private frame = 0;
  private lifeTime = 0;
  private temporary = false;
  private type: CoinType = CoinType.STATIC;
  private collides = false;

  static spawnDeathCoins(count: number, center: ReadonlyVec2, lifeTime: number, room: Room): void {
    for (let i = 0; i < count; i++) {
      const vel = sincos(Math.PI * (i + 1) / (count + 1));
      const velocity = new ImmutableVec2(vel.x * 100, vel.y * -200);
      const coin = new Coin();
      coin.setLifeTime(lifeTime);
      coin.setPosition(center);
      coin.setVelocity(velocity);
      room.addEntity(coin);
    }
  }

  constructor() {
    super();
    this.animationCoin = Resource.getAnimation('data/images/coin.bmp', 4);
    this.setSize(new ImmutableVec2(COIN_SIZE, COIN_SIZE));
  }

  getLayer(): number { return 3; }

  setLifeTime(lt: number): void {
    this.lifeTime = lt;
    this.type = CoinType.DYNAMIC;
    this.temporary = lt !== 0;
  }

  setCollides(): void { this.collides = true; }
  isDynamic(): boolean { return this.type === CoinType.DYNAMIC; }

  override onRespawn(): void {
    if (this.type === CoinType.DYNAMIC) this.remove();
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    let x = this.getPosition().x - this.animationCoin.getFrameWidth() / 2 + offsetX;
    let y = this.getPosition().y - this.animationCoin.getFrameHeight() / 2 + offsetY;

    if (!this.temporary || Math.floor(this.lifeTime / 10) % 2 === 0) {
      let f = this.frame;
      if (f > 180) f = 0;
      this.animationCoin.drawFrame(ctx, (f < 5 * 4) ? Math.floor(f / 5) : 0,
        Math.round(x), Math.round(y), false, false);
    }
  }

  override update(): void {
    if (this.type === CoinType.DYNAMIC) {
      if (this.myRoom.getHero().Collides(this)) {
        this.setCollides();
      }

      if (this.temporary && this.lifeTime === 0) {
        this.remove();
      } else {
        this.myVelocity.add(0, 6.0);
        const bumps = this.moveWithCollision();

        if (bumps.has(Direction.LEFT) || bumps.has(Direction.RIGHT)) {
          if (Math.abs(this.myVelocity.x) > 10) Sound.playSample('data/sounds/coin');
          this.myVelocity.set(this.myVelocity.x * -0.5, this.myVelocity.y);
        }
        if (bumps.has(Direction.UP) || bumps.has(Direction.DOWN)) {
          if (Math.abs(this.myVelocity.y) > 10) Sound.playSample('data/sounds/coin');
          this.myVelocity.set(this.myVelocity.x, this.myVelocity.y * -0.2);
        }
        this.lifeTime--;
      }
    }

    if (this.collides) {
      if (this.myRoom.getHero().gotCoin()) {
        PlayerSkill.playerDidSomethingClever(0.3, 0.05);
        Sound.playSample('data/sounds/coin');
        this.remove();
      }
    }

    this.frame++;
  }
}

File: wii/src/entities/coin.cpp
#include "entities/coin.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "room.h"
#include "player_skill.h"
#include "utils.h"
#include <cmath>

Coin::Coin() {
    animationCoin_ = Resource::getAnimation("data/images/coin.bmp", 4);
    setSize(Vec2(12, 12));
}

void Coin::spawnDeathCoins(int count, Vec2 center, int lifeTime, Room* room) {
    for (int i = 0; i < count; i++) {
        Vec2 vel = sincos(static_cast<float>(M_PI) * (i + 1) / (count + 1));
        Vec2 velocity(vel.x * 100.0f, vel.y * -200.0f);
        Coin* coin = new Coin();
        coin->setLifeTime(lifeTime);
        coin->setPosition(center);
        coin->setVelocity(velocity);
        room->addEntity(coin);
    }
}

void Coin::setLifeTime(int lt) {
    lifeTime_ = lt;
    type_ = CoinType::DYNAMIC;
    temporary_ = (lt != 0);
}

void Coin::setCollides() {
    collides_ = true;
}

void Coin::onRespawn() {
    if (type_ == CoinType::DYNAMIC) remove();
}

void Coin::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(position_.x)) - animationCoin_->getFrameWidth() / 2 + offsetX;
    int y = static_cast<int>(std::round(position_.y)) - animationCoin_->getFrameHeight() / 2 + offsetY;

    if (!temporary_ || (lifeTime_ / 10) % 2 == 0) {
        int f = frame_;
        if (f > 180) f = 0;
        int frameIdx = (f < 5 * 4) ? (f / 5) : 0;
        animationCoin_->drawFrame(renderer, frameIdx, x, y);
    }
}

void Coin::update() {
    if (type_ == CoinType::DYNAMIC) {
        if (room_->getHero()->Collides(*this)) {
            setCollides();
        }

        if (temporary_ && lifeTime_ == 0) {
            remove();
        } else {
            velocity_.y += 6.0f;
            auto bumps = moveWithCollision();

            if (bumps.count(Direction::LEFT) || bumps.count(Direction::RIGHT)) {
                if (std::abs(velocity_.x) > 10.0f) Sound::playSample("data/sounds/coin");
                velocity_.x *= -0.5f;
            }
            if (bumps.count(Direction::UP) || bumps.count(Direction::DOWN)) {
                if (std::abs(velocity_.y) > 10.0f) Sound::playSample("data/sounds/coin");
                velocity_.y *= -0.2f;
            }
            lifeTime_--;
        }
    }

    if (collides_) {
        if (room_->getHero()->gotCoin()) {
            PlayerSkill::playerDidSomethingClever(0.3f, 0.05f);
            Sound::playSample("data/sounds/coin");
            remove();
        }
    }

    frame_++;
}
